protocol/binary/stream_reader.go

```diff
@@ -66,24 +66,51 @@ type StreamReader struct {
 
 	reader io.Reader
 	buffer [8]byte
+
+	// discard is points to either discardOffset or discardStream based on
+	// the implementation of the io.Reader we're using.
+	discard func(int64) error
+
+	// These are bound versions of the discardStream and discardOffset
+	// methods on the StreamReader. Putting them here ensures that we don't
+	// cause an alloc when we do "sr.discard = sr.discardOffset".
+	_discardStream func(int64) error
+	_discardOffset func(int64) error
+
+	// This field is set only if the wrapped reader is an offset reader.
+	// DO NOT USE if you're not discardOffset.
```

Comment:
really nitty nit: avoid double negative - "ONLY USE if you are ..."
```suggestion
	// ONLY USE if you are discardOffset.
```